### DISCUSSION TITLE
Trouble with ERG Mode

### DISCUSSION THREAD

**Discussion by Lunchtime0614** [doudar/SmartSpin2k#344] [Q&A] [answered]
I just built this as a kit from @eMadman (many thanks, put together beautifully, and I guess you can ignore my email on this - might be good for the public to see the issue here)...I'm using it with Assioma Duo pedals and have tried a few different structured workout apps...and I can't seem to get it to work right.

My main app is Trainer Day (not trainer road) and it doesn't control the SS2K - for example if the app calls for a 120 watts it doesn't turn the resistance up nearly close enough - stays around 35-50 watts.

I tried Kinomap - it adjusts the resistance but way high - if it calls for 70 watts it was maxed out on the resistance knob of the bike.

Looking at the debug screen (during Kinomap) I see something like FTMS Server - Target watts 256, current watts 0.  Even though in the current watts section of the debug screen it shows the actual correct watts.

On the debug screen (during Trainer Day)  I see 'Control Request', and ERG Mode Target: 1 while up above it shows my current watts from the pedals which is like 20...makes sense if it's trying to go to 1.

**Comment by MarkusSchneider** [comment]
Can you attach a log. I also use Assioma DUO pedals and it works for me.
Can you see your power meter on the debug page?
Can you see power + cadence?

**Reply by Lunchtime0614** [reply]
I'm not that great at the software coding side of things - how do I install logbert?  I went to the Github page for it as referenced in the SS2K wiki but I'm not seeing how to install it?

**Reply by MarkusSchneider** [reply]
You can copy the messages from the debug page.
If you want to install logbert got to the download page https://github.com/couchcoding/Logbert/releases
and download the msi package.
But you must then configure the UPD Appender in Logbert and enable UPD Logging in SS2k.

**Reply by Lunchtime0614** [reply]
Here's two logs from Logbert - One for TrainerDay, one for Kinomap.  Trainer Day when on the debug screen sets the target watts to 1 app screen calls for 120, Kinomap sets it to 256 immedially even though the app screen  is calling for 50.

[TrainerDay_UDP.csv](https://github.com/doudar/SmartSpin2k/files/8984863/TrainerDay_UDP.csv)

[KinomapUDP.csv](https://github.com/doudar/SmartSpin2k/files/8984864/KinomapUDP.csv)

TrainerDay showing 120w but only being set to 30-50 from SS2K
![TrainerDaySCREEN_120Watt](https://user-images.githubusercontent.com/68966864/175776586-f561cc54-1da6-4b1c-90d4-bc7ba1853b07.jpg)

Kinomap Debug Screen (sorry I forgot to capture the Trainerday Debugscreen but its about the same just with 1 as the target watt value)
![KinoMapDEBUGSCREEN](https://user-images.githubusercontent.com/68966864/175776696-664498f8-595b-4de6-817c-3b87d9eeded3.JPG)


Kinomap App Screen calling for 50w
![KinoMapSCREEN-50Watt](https://user-images.githubusercontent.com/68966864/175776672-0027f325-2d64-41f9-86ff-9cd5eb394126.JPG)

**Reply by Lunchtime0614** [reply]
Oh and Kinomap is from an iPhone 13. Trainer Day as shown here is on Android...but it does the same when used on the iPhone.

**Comment by doudar** [comment]
There's definitely something going on with the FTMS communication in those logs. Are you able to test any other apps in the meantime?

**Reply by Lunchtime0614** [reply]
Sure - what app would you like me to try.  For what it's worth, I also have the Shift Smart Trainer (a beta version of it - I think that project is defunct because they quit communicating)...and it had this same problem with Trainer ROAD - same problem of not increasing the resistance on the wheel.

**Comment by MarkusSchneider** [comment]
What I can see from the TrainerDay log is that only 50 Watts are requested form trainer.
The value from the power meter is 49 Watts.
ERG seems to be OK.

**Reply by doudar** [reply]
> What I can see from the TrainerDay log is that only 50 Watts are requested form trainer. The value from the power meter is 49 Watts. ERG seems to be OK.

I think you're seeing 50 because it's the default minimum control value. I don't see the actual value of 50 being sent to SS2K anywhere. 

My initial thought is that both of those apps don't like somthing in Fitnessmachinefeature or fitnessmachinestatus

**Reply by MarkusSchneider** [reply]
Yes you are right. values sent are around 30 Watts

**Reply by Lunchtime0614** [reply]
Anything specific you want me to test... other than a different app as @doudar  suggested (just let me know what app - I was going to try trainer road but I don't have a subscription anymore).

**Comment by Lunchtime0614** [comment]
@doudar and @MarkusSchneider  - Here is zwift...from my iphone.  Seems to be working fine.  For reference - I assume this is the version of the firmware (shown on the default page of SS2K via the web) - 2.2.8-develop-23-gce1ad14


[zwift_IphoneUDP.csv](https://github.com/doudar/SmartSpin2k/files/8985194/zwift_IphoneUDP.csv)


Here's a video of the debug screen when Zwift is running - as I'm starting up....

https://drive.google.com/file/d/1kemZHAG1bFUTPb4mL2ifCGojI8VQHRrT/view?usp=sharing

**Reply by doudar** [reply]
Well that's good news. Sounds like an app compatibility issue. I'll start working on it (unless @MarkusSchneider beats me to it :) ). Thanks for reporting it.

**Reply by Lunchtime0614** [reply]
FYI - Trainer Day is a relatively new app - and I've spoken to the developer about it for an unrelated issue...He might be open to lending a hand - trainerday.com or forums.trainerday.com where he's most responsive - (Alex is his name).  I have a direct message chain with him from a while ago I'll point him to this thread.  No harm I suspect.

**Reply by Lunchtime0614** [reply]
> Well that's good news. Sounds like an app compatibility issue. I'll start working on it (unless @MarkusSchneider beats me to it :) ). Thanks for reporting it.

Does it seem odd that both TrainerDay and Kinomap have an issue - oddly enough opposite but related issues?

**Reply by avanlaningham** [reply]
@htims05 So you said Zwift worked.  To confirm you mean ERG mode in Zwift? watching the log it looks like it is ERG.  Alex from TrainerDay here.  Yes this project is interesting.  The fist I have heard of it.  We support standard FTMS for ERG, Resistance and Slope modes.  It works with all smart trainers that work via FTMS that we know of.  Not sure how I can help here. @htims05 If you have an Android device it makes debugging on our side much easier, but we can try debugging in iOS too.  We have a new test mode we can enable on your phone.

**Comment by Lunchtime0614** [comment]
@doudar and @MarkusSchneider  - Alex from Trainer Day just messaged me and said he's very interested in this and would like to help out.  I have his email address but I don't know how to add him to this thread without publicly putting out his email?

**Reply by doudar** [reply]
Just send him the link to this GitHub discussion

**Reply by Lunchtime0614** [reply]
I did - I guess if he want's to be on the thread he'll have to register and jump in on the thread.  That said he did read through this and asked me to try resistance mode and adjust with the up/down buttons on the app.  I did that and when I switched to resistance the SS2K just went full resistance until the motor started slipping as the resistance knob was locked out.  As soon as I went back to ERG it went down to 50 again.  Tried on both iphone and android...same result.

**Reply by doudar** [reply]
Testing Kinomap - the SS2K is responding to the instructions from the app correctly. At the lowest level, I'm seeing this:

D NimBLECharacteristic: "Characteristic 0x2ad9 Write event"
D NimBLECharacteristic: ">> setValue: length=3, data=050001, characteristic UUID=0x2ad9"
for any setpoints less than 256. 05 is watt control, 0001 is the setpoint in LE hex (256). 

If you increase the setpoint above 256, you start to see 05 04 01 (the second octet finally changing).   

I'm guessing what we're seeing is the result of Kinomap not error handling (an incorrect) setting in our fitness machine feature characteristic properly.

**Reply by doudar** [reply]
I found the issue. These must be the first two apps that we've run into that actually use the ftmsPowerRange Characteristic. 
In that setting we had our endianness confused. 

was:
uint8_t ftmsPowerRange[6]  = {0x00, 0x01, 0xA0, 0x0F, 0x01, 0x00}; 

fixed:
uint8_t ftmsPowerRange[6]  = {0x01, 0x00, 0xA0, 0x0F, 0x01, 0x00}; 

This was also transposed the same on:
ftmsResistanceLevelRange[6]      = {0x01, 0x00, 0x25, 0x00, 0x01, 0x00};

So I'll assume that was the issue you saw with resistance level as well. 

@htims05 , I'll have a binary for you to test within minutes.

**Reply by doudar** [reply]
[firmware.zip](https://github.com/doudar/SmartSpin2k/files/8985470/firmware.zip)

@htims05 go ahead and try this binary. You can upload it via the web interface if you'd like. Just extract it and make sure it's named exactly firmware.bin
user and pass are both admin.

It's working on my end with kinomap, I'll try TrainerDay in a minute. 

Good luck!

**Comment by doudar** [comment]
Confirmed working with TrainerDay as well.

**Reply by avanlaningham** [reply]
Oh cool :) I will read into this project more after I get home from vcation.  Super cool.  Have you heard of anything for anyone with a lever rather than knob?  I have a Keiser M3i.  I also have many smart trainers but it would be cool to try it on the M3i.

**Reply by doudar** [reply]
I'm sure you could get it working with some extreme hardware modifications :)

Btw, congratulations on following the BLE spec so well. It turns out your app was doing exactly what the SS2K was telling it to and not setting any power less than 256w :)

**Comment by doudar** [comment, marked as answer]
Doubling this here so it's not buried:

[firmware.zip](https://github.com/doudar/SmartSpin2k/files/8985481/firmware.zip)


@htims05 go ahead and try this binary. You can upload it via the web interface if you'd like. Just extract it and make sure it's named exactly firmware.bin
user and pass are both admin.

This binary also has the latest updates to LittleFS so it should be much more stable in general. 

Once we get some test results in, I'll turn this into an OTA release.

**Reply by Lunchtime0614** [reply]
I'll give this a go tonight...actually out mountain biking right now!!

**Reply by Lunchtime0614** [reply]
@doudar  - Does this mean it's done....doesn't refresh or give any message that it's done?  I'm not sure if it's just 100% uploaded or actually 100% updated and complete.

![Screen Shot 2022-06-25 at 7 37 22 PM](https://user-images.githubusercontent.com/68966864/175793638-db48d84e-7dbd-4f26-800f-a1bac667d3de.png)

**Reply by doudar** [reply]
Yep, if it hits 100%, it's done.

**Reply by doudar** [reply]
I should add that if you don't currently have it connected to the internet, after that update you'll have to enter your WiFi information so it can download the web UI files from GitHub.

**Reply by Lunchtime0614** [reply]
@doudar  - the new firmware fixes erg mode in Trainer Day, but now I'm not getting my HR passed through the SS2K.  I thought perhaps the new firmware wiped out the memory and I had to re-pair it, so I turned it off, did a scan so it saw none, saved, turned HR back on, Scan, and saved.  Debug screen doesn't show HR, nor does Trainer Day like before.  I haven't had the HR connected to anything else since earlier today.

**Reply by doudar** [reply]
Nothing changed there so all I can recommend is to make sure nothing else is grabbing your HR first (turn off BLE on your phone if needed) until it's completed.

**Reply by Lunchtime0614** [reply]
I'll do some testing with it tomorrow....and report back.  I also have my scheduled training ride for Monday so I'll let you know how that goes.

**Reply by Lunchtime0614** [reply]
Ok Got it to work today, I'm not sure if it was because I connected a different HRM and then back to this one or if it finally released from whatever it was holding on to even though those devices had bluetooth off.

One thing I did notice is that the HRM doesn't auto connect to SS2K if the HRM is already on before the SS2k - like the assioma pedals do.

Full ride test tomorrow with Trainer Day.

**Reply by Lunchtime0614** [reply]
@doudar, @MarkusSchneider - the new firmware seems to have solved the issue...and I'll say that the SS2k held the target power in ERG mode much better than anything I've ever tried before.  The shift smart trainer was never anywhere close to being this consistent.  I ran this on Trainer Day with a 6s power smoothing, @avanlaningham  this works fantastically!!

The only thing I think I saw that might have been an issue (more testing needed) is in the later half of the ride on the recoveries if I increased cadence it didn't seem to drop the resistance down to to the target.   Now I realized after the ride that I left the debugging screen open on a different computer, I'm by no means knowledgable here but I wonder if it was caching the log and it got very large causing it to be less responsive.  You can see how it became jumpy in the recoveries at the end of the ride. (Don't mind the shift in the target power line, I accidentally skipped forward (read: shortened) a recovery and I guess intervals doesn't recognize that shift).

Ride with SS2K:
![Screen Shot 2022-06-27 at 3 35 56 PM](https://user-images.githubusercontent.com/68966864/176022786-27214e42-ec78-42df-99b3-e454fe5d5439.png)


Ride with Shift Smart Trainer:
![Screen Shot 2022-06-27 at 3 44 33 PM](https://user-images.githubusercontent.com/68966864/176023243-186df999-8be8-4049-940e-62be5286f856.png)

**Reply by doudar** [reply]
@htims05 thats great! Thanks for the report back.

@MarkusSchneider and I spent a bunch of time this winter tuning the ERG code, so it makes me pretty happy to know it's working well for you. 

As far as the high to low cadence, you could try playing around with the ERG sensitivity and see if it helps. 

The way ERG currently works is that anytime you are pedaling over 50rpm, SS2K is creating and updating a power vs resistance lookup table. Then when you are in ERG mode and a new power target is sent, it uses that table to interpolate between the two nearest resistance settings that match the requested target. It then spins the knob rapidly to that setting. It then waits a couple seconds for the power to stabilize. Then it goes into the classic ERG mode where it uses the   erg sensitivity in a closed feedback loop to try to maintain the power target. It stays in that mode until a new power target is sent. 

All that said, there's an assumption built into the powertable lookup algorithm to convert different cadences which (if I assumed wrong ;) ) could throw off the initial resistance that the knob spins to after a new power target is sent.

**Reply by Lunchtime0614** [reply]
Awesome!  Fantastic work...one last question - albeit not related to this thread.  Are HRM's supposed to auto connect if they are on when the SS2K is powered up?  Mine never does, seems that I always have to go hit the scan button and then save before it picks it up.  Just hitting scan doesn't make it connect even though the name is already listed in the dropdown.  

My assioma pedals are always connected and pushing values to the debug screen as soon as I turn on the SS2K.

**Reply by doudar** [reply]
It *should* just auto connect every time...

You could try just leaving it in "any" and see it that works. "Any" should work fine as long as there aren't any other HRM's within range.
